### DISCUSSION TITLE
GLPI Inventory Tasks of GLPI Agent

### DISCUSSION THREAD

**Discussion by dethorpe** [glpi-project/glpi-agent#390] [Q&A]
Hi all, I hope I am posting this in an appropriate place. I seem to be missing some of the connective tissue between getting the GLPI server and GLPI agent to work together for network and esx inventories.

I believe my setup is 90% there, the GLPI server is running and accepting regular inventory from glpi-agents. glpi-agents have the appropriate modules installed and I can run manual esx inventories by hand (IE #glpi-ex --host xxx.xxx.x.xxx --user foo --password foo2 --directory )

I have created a task on the server with those details above that I "think" should be triggering the agent to do it's thing however that never seems to leave "prepared" and "not done yet" if I am understanding the Job execution information. I also seem to have a blind spot in the documentation here for the interrelationships between the server and agent.

If someone has a minute to help me diagnose, I'd be happy to pay it forward with some time writing up the experience for the documentation.

Thanks a million.

(cross post from https://github.com/glpi-project/doc-agent/issues/4)

**Comment by g-bougard** [comment]
Hi @dethorpe 

first is you want to pay for help, the best way is to [earn a GLPI subscription to request professional support](https://glpi-project.org/prices/). To do so, you can directly contact one of Editor's partners, see: https://glpi-project.org/partners/

Speaking about your question, if your task is not prepared, there can be few reasons for that:
1. you missed to enable the crontab for GLPI and then automatic actions are not started. You should check the journal of automatic actions if they are really started.
2. you missed to enable the related task on the agent configuration in GLPI
3. you missed some configuration on the task
4. you forget to enable the "Permit to re-prepare task after run" on the task
5. you missed to create a VMWare Host credential and to associate it to you "Remote device inventory"

**Reply by dethorpe** [reply]
Thanks for your reply and thank you for moving this to the appropriate forum,

> you want to pay for help

I am aware of paid support and this is indeed a good option. However when I say "pay it forward" I am referencing a colloquial in English that means: "do a favor for some stranger in the future for a favor done for you" which I say because I would prefer to contribute to improving a pathway for others to help themselves then buy an answer only for myself.

>you missed to enable the crontab for GLPI and then automatic actions are not started. You should check the journal of automatic actions if they are really started.

I may have, however I have run through the automatic actions and manually triggered an execution of any action that appears related to agent, inventory, or task with no positive result. Were this to work, I would of course schedule the actions to run via chronjob but I think I misunderstand the suggestion or this is not the culprit.

>you missed to enable the related task on the agent configuration in GLPI

I think I did this. Under Task Management of the task it is marked "active." Tangentially, I have also triggered the Force Start action here as well.

>you missed some configuration on the task

I'm not sure what... it has a scheduled start date, a preparation timeslot and execution timeslot that covers all time. It can "wakeup" 100 agents every minute. It has Targets (the esx item to be inventoried?) and Actors (the Agents to inventory?)

>you forget to enable the "Permit to re-prepare task after run" on the task

I did not have this checked, however the verbiage suggests to me this is a post execution option and it has not executed once.

>you missed to create a VMWare Host credential and to associate it to you "Remote device inventory"

I don't believe I did. That is setup in GLPI and those creds are tested on the agent CLI.

Like I said, I feel like I am just missing something simple.

**Reply by g-bougard** [reply]
> > you missed to enable the related task on the agent configuration in GLPI
> 
> I think I did this. Under Task Management of the task it is marked "active." Tangentially, I have also triggered the Force Start action here as well.

As you said, this is probably a simple misunderstanding and here there's one. Agents are listed in Administration > Inventory > Agents (at the top). There, edit the agent you selected as "actor". Open the "Agent modules" tab and be sure the "VMware host remote inventory" task option is checked.

**Reply by dethorpe** [reply]
Thank you for your replay. I am certain that is correct.
<img width="226" alt="image" src="https://github.com/glpi-project/glpi-agent/assets/406266/ef1c265e-9789-4595-b856-a041ada0b805">

Also please note:
<img width="555" alt="image" src="https://github.com/glpi-project/glpi-agent/assets/406266/2d39b183-0315-4d74-8264-fa787d021faa">
<img width="449" alt="image" src="https://github.com/glpi-project/glpi-agent/assets/406266/0b85ee43-5520-4e82-842e-f6801a1cacda">
<img width="346" alt="image" src="https://github.com/glpi-project/glpi-agent/assets/406266/651994ed-e3fb-4ddf-8040-582186713997">
<img width="601" alt="image" src="https://github.com/glpi-project/glpi-agent/assets/406266/9728fcb1-3835-4e53-b70e-5c4c5925cd5d">

**Reply by g-bougard** [reply]
> Thank you for your replay. I am certain that is correct.
> <img alt="image" width="226" src="https://user-images.githubusercontent.com/406266/237467671-ef1c265e-9789-4595-b856-a041ada0b805.png">

Here there's still a misunderstanding. "Enabled tasks" is just a report of what the agent reported to GLPI which tasks it supports. So this means the agent supports ESX task and so your agent is well installed. That's still a good point.

But read me again: >>> edit the agent you selected as "actor". Open the "Agent modules" tab and be sure the "VMware host remote inventory" task option is checked.

So on the agent, you really need to select the "Agent modules" tab, don't stop your reading on the "Enabled tasks" frame which is not what I was speaking about. And indeed, as I was confused like you and for some other reasons, I asked GLPI team to modify the agent page so "Enabled tasks" will be replaced by "Supported tasks" (see glpi-project/glpi#14692) for next GLPI release.

This is what I was talking about:
![image](https://github.com/glpi-project/glpi-agent/assets/12514489/b9b2b997-dc14-4a74-82ba-57cd771ecee0)

**Reply by dethorpe** [reply]
>But read me again: >>> edit the agent you selected as "actor". Open the "Agent modules" tab and be sure the "VMware host remote inventory" task option is checked.

Apologies! I attested just the agent status as reported to GLPI. But also GLPI's enabled Agent modules also is correct:

<img width="162" alt="image" src="https://github.com/glpi-project/glpi-agent/assets/406266/8a8476dc-5e46-4acc-8fdd-349b0fd5338c">

>And indeed, as I was confused like you and for some other reasons... see https://github.com/glpi-project/glpi/pull/14692

Thank you! It is great to see!

**Reply by g-bougard** [reply]
Okay nice.

During manual tsting, you should not set "Schedule start" and Preparation/Execution timeslots on the task in GLPI. Then you have to "Force start" the task in GLPI and manage to force the agent. To force the agent, you should access `http://localhost:62354` url from a browser on the agent computer or request `http://localhost:62354/now` with curl or wget if this is an unix/linux server with only a ssh access. After that you can check the task log in glpi and the agent log to see what happens.

**Reply by dethorpe** [reply]
Thank you for sticking with me!

>During manual tsting, you should not set "Schedule start" and Preparation/Execution timeslots on the task in GLPI.

LOL, I did not initially. But added it along my own troubleshooting path. I removed them now.

>Then you have to "Force start" the task in GLPI and manage to force the agent.

From that Task Management panel. Not some other GLPI automatic action, I assume. 

> To force the agent, you should access http://localhost:62354

In the days of fusion this control used to have a small url to force the inventory. I no longer see it with the GLPI agent. However navigating to the /now domain path via browser or curl return "Access Denied" status from the Agent (not 403.) I do not see any error reported in the glpi-agent.log though.

**Reply by dethorpe** [reply]
Looking into https://github.com/glpi-project/glpi-agent/issues/42 and I might be able to guess some questions you'll ask!
Ubuntu 20.04, agent 1.4-1, nothing much in the config file except the target server (which in this case it actually itself.) Anyhow the self inventory works fine regularly.

**Reply by dethorpe** [reply]
overriding the logging level to debug 1 I see: "invalid request (untrusted address)" however I don't see how to add scope a trusted IP range.

**Reply by dethorpe** [reply]
Here is an interesting bit of logs, especially that last "remote inventory not supported"

`[Thu May 11 09:44:42 2023][debug] target server0: init event for ESX task`
`[Thu May 11 09:44:42 2023][debug] target server0: init event for RemoteInventory task`
`[Thu May 11 09:44:42 2023][debug] target server0: init event for NetDiscovery task`
`[Thu May 11 09:44:42 2023][debug] target server0: init event for NetInventory task`
`[Thu May 11 09:44:43 2023][debug] target server0: init event for Collect task`
`[Thu May 11 09:44:43 2023][debug] target server0: init event for WakeOnLan task`
`[Thu May 11 09:46:41 2023][debug] target server0: storage maintenance event for deploy task`
`[Thu May 11 09:46:41 2023][debug] forking process 876311 to handle child job`
`[Thu May 11 09:46:41 2023][debug] forking process 876311 to handle task Deploy`
`[Thu May 11 09:46:41 2023][info] running task Deploy: storage maintenance event`
`[Thu May 11 09:46:41 2023][debug] Deploy task storage maintenance event for server0 target`
`[Thu May 11 09:46:41 2023][debug] No need to plan another storage maintenance event for server0 target`
`[Thu May 11 10:00:30 2023][info] target server0: server http://xxxx/front/inventory.php`
`[Thu May 11 10:00:30 2023][debug] [http client] 4DA1FE5C: Using Compress::Zlib for compression`
`[Thu May 11 10:00:30 2023][info] sending contact request to server0`
`[Thu May 11 10:00:30 2023][debug] server message: remoteinventory task not supported`

As far as I can tell it should be enabled and supported... see previous... right?

**Reply by g-bougard** [reply]
Hi @dethorpe 
about your "Access denied" problem, you should add "127.0.0.1" to the agent `httpd-trust` parameter.
About "remote inventory not supported", don't be confused, ESX task is not remoteinventory task in the agent context and as I see "target server0: init event for ESX task", it means ESX task is enabled. Actually, this message only means there's still no support on GLPI side to provide "remoteinventory" task configuration.
RemoteInventory task is to inventory other platform via SSH for Unix/Linux & via WinRM for windows (see https://glpi-agent.readthedocs.io/en/latest/tasks/remote-inventory.html).

Finally, I see you're using `http://xxxx/front/inventory.php` URL... It is wrong. Take car of the caution notes here: https://glpi-agent.readthedocs.io/en/latest/configuration.html#server
Your URL should be `http://xxxx/marketplace/glpiinventory/` or `http://xxxx/plugins/glpiinventory/` depending on the way you installed the plugin.

**Reply by dethorpe** [reply]
>about your "Access denied" problem, you should add "127.0.0.1" to the agent httpd-trust parameter.

That did it!

>>[Thu May 11 10:00:30 2023][info] target server0: server http://xxxx/front/inventory.php

>Finally, I see you're using http://xxxx/front/inventory.php URL... It is wrong. Take car of the caution notes here: https://glpi-agent.readthedocs.io/en/latest/configuration.html#server

Thanks for the notes. I corrected that although the address seems to be a default value, as I don't see it explicit anywhere except a comment in agent.cfg
`$ grep front /etc/glpi-agent/*`
`/etc/glpi-agent/agent.cfg:#server = http://server.domain.com/front/inventory.php`
Maybe this is worth patching correct in future versions...
I think this is tangental though, I do not think this is causing a problem (especially not the one I am trying to solve) as inventory of the system this agent is installed on reaches GLPI fine but it's ability to inventory remote systems through a task still does not appear to be working.
Where else can we look?
`[Fri May 12 12:43:41 2023][info] running task ESX`
`[Fri May 12 12:43:41 2023][error] [http client] communication error: 404 not found `
Could this sequince be indicating an error?

**Reply by g-bougard** [reply]
I just updated the default config for next nightly to include another sample:
https://github.com/glpi-project/glpi-agent/commit/336364280bbc2d65e0dd7763d14eaa4bb0b99409

About you 404 not found... something is probably still wrong in your URL or in your server. You should set `debug = 2` in agent config to have a complete debug of what is requested in the agent. For ESX task, this also can be an ESX url issue.

**Reply by dethorpe** [reply]
At debug level 2, I've had to turn inventory off to just filter down the signal from noise. I hope that isn't a problem. I think the problem is being described here:
`[Mon May 15 01:15:33 2023][info] running task ESX`
`[Mon May 15 01:15:33 2023][debug2] http://DNSNAME/marketplace/glpiinventory/?action=getConfig&task[ESX]=2.7&machineid=GLPI-23-2023-05-04-19-35-07`
`[Mon May 15 01:15:33 2023][debug2] http://xx.xxx.xxx.xx/glpi/marketplace/glpiinventory/b/esx/?action=getJobs&machineid=GLPI-23-2023-05-04-19-35-07`
`[Mon May 15 01:15:33 2023][error] [http client] communication error: 404 Not Found`
It strikes me as significant that I define the server by hostname in my config yet it appears to have picked up the ip address here for some reason. Either should work but the second has added a to the path in a way that I do not think exhists. Where is it getting this second url? It certain isn't something I have defined.

Me thinks a good reference for someone else someday.
I triggered the job by hand, not just by "force start" but also "restart selected jobs"
<img width="184" alt="image" src="https://github.com/glpi-project/glpi-agent/assets/406266/38d3faf8-27f2-48c2-b2e6-a77f2903bc58">

**Reply by g-bougard** [reply]
Now you probably have to fix your "Agent Base URL" on your entity in the "Assets" tab:
![image](https://github.com/glpi-project/glpi-agent/assets/12514489/197c41a4-8879-4588-b1f3-afc1ad34b68f)
If you have an url there, first try to set it empty to leave GLPI Agent to use the default one.

**Reply by dethorpe** [reply]
We did it!
Thank you.
